### PR TITLE
add non-channel query iterator method

### PR DESCRIPTION
**Pull request by kevina**
Continuation of #53

### CODE REVIEW

**Comment by whyrusleeping** [comment]
@Kubuxu could you help CR this one?

**Review by whyrusleeping — APPROVED** [review]
Overall LGTM, the 'having to switch back to legacy mode' kinda sucks. Maybe theres a better way of doing that? Not a huge deal though.

**Comment by whyrusleeping** [comment]
Maybe on conversion to legacy we swap out the next function in case the user calls `NextSync` again?

**Comment by kevina** [comment]
@whyrusleeping I confused, what are you getting at.  If `Next()` is called and then `NextSync()` do you want the `NextSync()` method to go through legacyResults instead?  Or do you want NextSync() to error out?

**Comment by whyrusleeping** [comment]
Lets just have `NextSync` go through the legacy method instead. That way we don't accidentally miss an entry thats trying to be sent on the legacy channel.

**Comment by kevina** [comment]
Okay will do.

**Comment by kevina** [comment]
@Kubuxu @whyrusleeping one thing I was unsure of is the handling of the automatic closing in NextSync().  Even if there is not a go process, there might still be something that needs to be closed, for example the leveldb iterator should probably be closed when done to help free up resources.  Comments/reviews on this aspect will be appreciated.  Also see https://github.com/ipfs/go-ds-leveldb/pull/1 for some context.

**Comment by whyrusleeping** [comment]
@kevina I'm thinking we should require that the user always call `Close` on the query when they are done with it.

File: query/query.go [hunk 1]
```diff
@@ -129,6 +130,11 @@ func (r *results) Next() <-chan Result {
 	return r.res
 }
 
+func (r *results) NextSync() (Result, bool) {
+	val, ok := <-r.res
+	return val, ok
```
Comment (kevina) [inline query/query.go]: Never mind, I was going off the comments you gave me in IRC.  If we require the user calls close all is well.

**Comment by kevina** [comment]
@whyrusleeping okay I cleaned up the logic of when legacyResults is used and when close() is called.

This still needs testing.

**Comment by whyrusleeping** [comment]
@kevina sounds good, ping me when you've added more tests  then

**Comment by kevina** [comment]
@whyrusleeping  Unit tests added. 

I will do a manual integration test tomorrow, but other than that it should be good to go.

**Comment by kevina** [comment]
> @kevina I'm thinking we should require that the user always call Close on the query when they are done with it.

From IRC:

> kevina: whyrusleeping: in order to match the interface it seams NextSync() should call the close function if it is defined
> whyrusleeping: okay, so if NextSync returns 'false' we assume things have been closed?
kevina: I think, that is what we have to do to match the interface.
> kevina: In go-ds-leveldb i.Release() will be called when it has no more results to send in runQuery(), see https://github.com/ipfs/go-ds-leveldb/blob/a37289361537b8109b1a4ec3c7f194c5d2b39414/datastore.go#L141
> kevina: when not using a channel the close function passes in will serve to call i.Release() see https://github.com/ipfs/go-ds-leveldb/blob/a37289361537b8109b1a4ec3c7f194c5d2b39414/datastore.go#L108
> kevina: basically I think I figured this out :)
> whyrusleeping: Okay, that makes sense

File: query/query.go [hunk 2]
```diff
@@ -248,9 +255,128 @@ func ResultsWithEntries(q Query, res []Entry) Results {
...
+			case <-worker.Closing(): // client told us to close early
+				return
+			}
+		}
+		return
+	})
+
+	go b.Process.CloseAfterChildren()
+
+	r.legacyResults = b.Results().(*results)
```
Comment (whyrusleeping) [inline query/query.go:381]: Is this assertion always safe?

Comment (kevina) [reply query/query.go:381]: Yes based on the code for ResultBuilder.Results() in the same file.

**Comment by kevina** [comment]
@whyrusleeping All tests pass when using this version with go-ipfs.  I would call this good to go.

**Comment by whyrusleeping** [comment]
@kevina Cool, Would just like to have @Kubuxu and/or @lgierth review this first

**Review by Kubuxu — APPROVED** [review]

**Pull request merged**